iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -132,15 +132,99 @@ class ForOpArgFolding final : public OpRewritePattern<scf::ForOp> {
   }
 };
 
+/// An ad-hoc pattern to convert scf.for loop-carried values from
+/// `vector<8xf16>` to `vector<4xf32>` by inserting `vector.bitcast` around
+/// scf.for boundaries.
+///
+/// Those loop-carried values will be lowered into SPIR-V local variables. This
+/// pattern allows packing f16 values into f32 variables tightly so that we can
+/// generate shader conformant SPIR-V.
+struct PackForOpInductionVarVector final : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter& rewriter) const override {
+    VectorType v8f16Type = VectorType::get({8}, rewriter.getF16Type());
+    VectorType v4f32Type = VectorType::get({4}, rewriter.getF32Type());
+
+    SmallVector<unsigned, 8> ivIndices;
+    for (auto it : llvm::enumerate(forOp.getRegionIterArgs())) {
+      if (it.value().getType() == v8f16Type) ivIndices.push_back(it.index());
+    }
+    if (ivIndices.empty()) return failure();
+
+    // Bit cast all init values from v8f16 to v4f32.
+    auto ivInitValues = llvm::to_vector<8>(forOp.getIterOperands());
+    for (unsigned index : ivIndices) {
+      Value oldValue = ivInitValues[index];
+      ivInitValues[index] = rewriter.create<vector::BitCastOp>(
+          oldValue.getLoc(), v4f32Type, oldValue);
+    }
```

Comment:
You could potentially merge those two loops above to simplify a bit the code.

Reply:
The above loop is for generating `ivIndices`, which is used by this loop's range. I feel this way is clearer. :)